addons/mail/static/src/models/message_sender/message_sender.js

```diff
@@ -0,0 +1,286 @@
+                    const regexp = new RegExp(
+                        '(\\s|^)(' + escapedSource + ')(?=\\s|$)',
+                        'g');
+                    htmlString = htmlString.replace(regexp, '$1' + emoji.unicode);
+                }
+            }
+            return htmlString;
+        }
+
+        async _postMailChannel(postData) {
+            let messageId;
+            const command = this.thread.composer._getCommandFromText(postData.body);
+            Object.assign(postData, {
+                subtype_xmlid: 'mail.mt_comment',
+            });
+            if (command) {
+                messageId = await this.async(() => this.env.models['mail.thread'].performRpcExecuteCommand({
+                    channelId: this.thread.id,
+                    command: command.name,
+                    postData,
+                }));
+            } else {
+                messageId = await this.async(() =>
+                    this.env.models['mail.thread'].performRpcMessagePost({
+                        postData,
+                        threadId: this.thread.id,
+                        threadModel: this.thread.model,
+                    })
+                );
+            }
+            return messageId;
+        }
+
+        async _postMailThread(postData) {
+            Object.assign(postData, {
+                subtype_xmlid: this.thread.composer.isLog ? 'mail.mt_note' : 'mail.mt_comment',
+            });
+            if (!this.thread.composer.isLog) {
+                postData.context = {
+                    mail_post_autofollow: true,
+                };
+            }
+            const messageId = await this.async(() =>
+                this.env.models['mail.thread'].performRpcMessagePost({
+                    postData,
+                    threadId: this.thread.id,
+                    threadModel: this.thread.model,
+                })
+            );
+            // FIXME: bus currently doesn't report that a message has been sent.
+            // It only works on channel. Therefore we need to update it manually.
+            const [messageData] = await this.async(() => this.env.services.rpc({
+                model: 'mail.message',
+                method: 'message_format',
+                args: [[messageId]],
+            }, { shadow: true }));
+            this.env.models['mail.message'].insert(Object.assign(
+                {},
+                this.env.models['mail.message'].convertData(messageData),
+                {
+                    originThread: [['insert', {
+                        id: this.thread.id,
+                        model: this.thread.model,
+                    }]],
+                })
+            );
+            this.thread.loadNewMessages();
+        }
+
+        /**
+         * Send a message to the server based on the thread model and handle RPC
+         * error.
+         * @private
+         * @param {Object} options
+         */
+        async _processMessagesToBeSent(options) {
```

Comment:
The callback situation is not quite right, if the message failed or if we are already in queue, this option param will be lost on the way.

The only way to make it right would be the have the "callback information" on the pending message itself.